thompson_sampling/thompson_sampling.py

```diff
@@ -38,29 +42,31 @@ def __init__(self,
 
     def run(self):
         results = []
-        epoch = 0.
+        epoch = 0.0
 
         with mp.Pool(processes=self.num_processors) as pool:
-            for p in range(self.num_processors):
-                r = pool.apply_async(self.fitness_function,
-                                     (self.strategy.sample(), ))
+            for _ in range(self.num_processors):
+                r = pool.apply_async(self.fitness_function, (self.strategy.sample(),))
                 results.append(r)
 
-            while epoch < self.epochs:
+            while epoch < self.epochs * self.num_processors:
                 r = results.pop(0)
                 if r.ready():
                     x = r.get()
                     self.strategy.update(*x)
-                    if 'on_update' in self.callbacks:
-                        if self.callbacks['on_update'](
-                                epoch, self.strategy,
-                                epoch == self.epochs - 1):
+                    if "on_update" in self.callbacks:
+                        epoch_tmp = np.round(epoch / self.num_processors, 5)
+                        if self.callbacks["on_update"](
+                            epoch_tmp,
+                            self.strategy,
+                            (self.epochs - 1 / self.num_processors) - epoch_tmp < 1e-5,
```

Comment:
use np.isclose(self.epochs - 1 / self.num_processors, epoch_tmp)

Reply:
fixed in ea43096ed4d8dd42aad9fd3ef9bf3d96a216b6ec